native/spark-expr/src/schema_adapter.rs

```diff
@@ -0,0 +1,376 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Custom schema adapter that uses Spark-compatible casts
+
+use crate::cast::cast_supported;
+use crate::{spark_cast, SparkCastOptions};
+use arrow_array::{new_null_array, Array, RecordBatch, RecordBatchOptions};
+use arrow_schema::{Schema, SchemaRef};
+use datafusion::datasource::schema_adapter::{SchemaAdapter, SchemaAdapterFactory, SchemaMapper};
+use datafusion_common::plan_err;
+use datafusion_expr::ColumnarValue;
+use std::sync::Arc;
+
+/// An implementation of DataFusion's `SchemaAdapterFactory` that uses a Spark-compatible
+/// `cast` implementation.
+#[derive(Clone, Debug)]
+pub struct SparkSchemaAdapterFactory {
+    /// Spark cast options
+    cast_options: SparkCastOptions,
+}
+
+impl SparkSchemaAdapterFactory {
+    pub fn new(options: SparkCastOptions) -> Self {
+        Self {
+            cast_options: options,
+        }
+    }
+}
+
+impl SchemaAdapterFactory for SparkSchemaAdapterFactory {
+    /// Create a new factory for mapping batches from a file schema to a table
+    /// schema.
+    ///
+    /// This is a convenience for [`DefaultSchemaAdapterFactory::create`] with
+    /// the same schema for both the projected table schema and the table
+    /// schema.
+    fn create(
+        &self,
+        required_schema: SchemaRef,
+        table_schema: SchemaRef,
+    ) -> Box<dyn SchemaAdapter> {
+        Box::new(SparkSchemaAdapter {
+            required_schema,
+            table_schema,
+            cast_options: self.cast_options.clone(),
+        })
+    }
+}
+
+/// This SchemaAdapter requires both the table schema and the projected table
+/// schema. See  [`SchemaMapping`] for more details
+#[derive(Clone, Debug)]
+pub struct SparkSchemaAdapter {
+    /// The schema for the table, projected to include only the fields being output (projected) by the
+    /// associated ParquetExec
+    required_schema: SchemaRef,
+    /// The entire table schema for the table we're using this to adapt.
+    ///
+    /// This is used to evaluate any filters pushed down into the scan
+    /// which may refer to columns that are not referred to anywhere
+    /// else in the plan.
+    table_schema: SchemaRef,
+    /// Spark cast options
+    cast_options: SparkCastOptions,
+}
+
+impl SchemaAdapter for SparkSchemaAdapter {
+    /// Map a column index in the table schema to a column index in a particular
+    /// file schema
+    ///
+    /// Panics if index is not in range for the table schema
+    fn map_column_index(&self, index: usize, file_schema: &Schema) -> Option<usize> {
+        let field = self.required_schema.field(index);
```

Comment:
Should it be `table_schema`? Looks like it maps from a column index in "table schema"? `required_schema` might be a projection on actual `table_schema`.

Reply:
So what is the difference between table_schema and required_schema? It looks like `required_schema` is a projection on `table_schema`, no?

Reply:
Well, yes, `required_schema` is the projection on the `table_schema`. But afaik `table_schema` is used only for predicate pushdown (via `SchemaMapper.map_partial_batch`). which does not use this field mapping.

Reply:
So the field order between `table_schema` and `required_schema` could be different, right? 

If the given `index` is based on `table_schema` (I suppose it is because the API states it is an field index in the table schema), the retrieved `field` may be incorrect:

```rust
let field = self.required_schema.field(index);
```

For example, `table_schema` has field `[ a, b, c]`, `required_schema` is `[b, c]`. Index 0 for `table_schema` should be column `a`, but above line gets `b`.

Reply:
For some additional context, this code was copied from DataFusion, and we renamed the field `projected_table_schema` to `required_schema` to better match Spark terminology.

We copied this code from DataFusion because we needed to change the `cast` to use our version.

Reply:
Hmm, looks like you're right. I thought this was being used by `SchemaMapper.map_batch` to get field_mappings, but it is not. Also, I don't see where this is used (either here or in DataFusion).

Reply:
Since it is copied, okay for now. Maybe this is a potential bug...